core/src/main/java/org/apache/accumulo/core/client/impl/TabletServerBatchWriter.java

```diff
@@ -699,7 +711,33 @@ else if (Tables.getTableState(context.getInstance(), table) == TableState.OFFLIN
 
     }
 
-    void addMutations(MutationSet mutationsToSend) {
+    void queueMutations(final MutationSet mutationsToSend) throws InterruptedException {
+      if (null == mutationsToSend)
+        return;
+      boolean transferred = queue.tryTransfer(new Runnable() {
+        final MutationSet m = mutationsToSend;
+
+        @Override
+        public void run() {
+          if (null != m) {
+            try {
+              addMutations(m);
+            } catch (Exception e) {
+              updateUnknownErrors("Error processing mutation set", e);
+            }
+          }
+        }
+      });
+      if (!transferred) {
+        try {
+          addMutations(mutationsToSend);
```

Comment:
Agreed. I believe that binning can be done concurrently. There is still a synchronization point later, after the binning, but I believe that is the right place to block. I updated this PR to make the Executor a daemon with a name.